imap_processing/mag/l1c/mag_l1c.py

```diff
@@ -39,19 +51,444 @@ def mag_l1c(
+
+    Parameters
+    ----------
+    epoch_data : numpy.ndarray
+        The existing timeline data, in the shape (n,).
+    gaps : numpy.ndarray
+        An array of gaps to fill, with shape (n, 2) where n is the number of gaps.
+        The gap is specified as (start, end) where start and end both exist in the
+        timeline already.
+
+    Returns
+    -------
+    numpy.ndarray
+        The new timeline, filled with the existing data and the generated gaps.
+    """
+    full_timeline: np.ndarray = np.zeros(0)
+
+    # When we have our gaps, generate the full timeline
+    last_gap = 0
+    for gap in gaps:
+        gap_start_index = np.where(epoch_data == gap[0])[0]
+        gap_end_index = np.where(epoch_data == gap[1])[0]
+        if gap_start_index.size != 1 or gap_end_index.size != 1:
+            raise ValueError("Gap start or end not found in input timeline")
+
+        full_timeline = np.concatenate(
+            (
+                full_timeline,
+                epoch_data[last_gap : gap_start_index[0]],
+                generate_missing_timestamps(gap),
+            )
+        )
+        last_gap = gap_end_index[0]
+
+    full_timeline = np.concatenate((full_timeline, epoch_data[last_gap:]))
+
+    return full_timeline
+
+
+def find_all_gaps(
+    epoch_data: np.ndarray, vectors_per_second_attr: Optional[str] = None
+) -> np.ndarray:
+    """
+    Find all the gaps in the epoch data.
+
+    If vectors_per_second_attr is provided, it will be used to find the gaps. Otherwise,
+    it will assume a nominal 1/2 second gap. A gap is defined as missing data from the
+    expected sequence as defined by vectors_per_second_attr.
+
+    Parameters
+    ----------
+    epoch_data : numpy.ndarray
+        The epoch data to find gaps in.
+    vectors_per_second_attr : str, optional
+        A string of the form "start:vecsec,start:vecsec" where start is the time in
+        seconds and vecsec is the number of vectors per second. This will be used to
+        find the gaps. If not provided, a 1/2 second gap is assumed.
+
+    Returns
+    -------
+    numpy.ndarray
+        An array of gaps with shape (n, 2) where n is the number of gaps. The gaps are
+        specified as (start, end) where start and end both exist in the timeline.
+    """
+    gaps: np.ndarray = np.zeros((0, 2))
+    if vectors_per_second_attr is not None and vectors_per_second_attr != "":
+        vecsec_segments = vectors_per_second_attr.split(",")
+        end_index = epoch_data.shape[0]
+        for vecsec_segment in reversed(vecsec_segments):
+            start_time, vecsec = vecsec_segment.split(":")
+            start_index = np.where(int(start_time) == epoch_data)[0][0]
+            gaps = np.concatenate(
+                (find_gaps(epoch_data[start_index : end_index + 1], int(vecsec)), gaps)
+            )
+            end_index = start_index
+    else:
+        # TODO: How to handle this case
+        gaps = find_gaps(epoch_data, 2)  # Assume half second gaps
```

Comment:
```suggestion
        gaps = find_gaps(epoch_data, MAG_RATE.TWO_VECS_PER_SEC)
```
If you have no normal mode data from which you can start the timeline/gap finding you can assume 2 vec/sec on both sensors starting at midnight and then try and fill that with any available BM